pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,484 @@
+		return propertyDiffResult(etf, eps, res)
+	case Update:
+		res := &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE}
+		return propertyDiffResult(etf, eps, res)
+	default:
+		return nil
+	}
+}
+
+func promoteToReplace(diff *pulumirpc.PropertyDiff) *pulumirpc.PropertyDiff {
+	kind := diff.GetKind()
+	switch kind {
+	case pulumirpc.PropertyDiff_ADD:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD_REPLACE}
+	case pulumirpc.PropertyDiff_DELETE:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE_REPLACE}
+	case pulumirpc.PropertyDiff_UPDATE:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE}
+	default:
+		return diff
+	}
+}
+
+func isForceNew(etf shim.Schema, eps *SchemaInfo) bool {
+	if etf != nil && etf.ForceNew() {
+		return true
+	}
+	if eps != nil && eps.ForceNew != nil && *eps.ForceNew {
+		return true
+	}
+	return false
+}
+
+func mapHasReplacements(m map[string]*pulumirpc.PropertyDiff) bool {
+	for _, diff := range m {
+		if diff.GetKind() == pulumirpc.PropertyDiff_ADD_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_DELETE_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_UPDATE_REPLACE {
+			return true
+		}
+	}
+	return false
+}
+
+func simplifyDiff(
+	diff map[string]*pulumirpc.PropertyDiff, key resource.PropertyKey, old, new resource.PropertyValue,
+	oldOk, newOk bool, etf interface{}, eps *SchemaInfo,
+) (map[string]*pulumirpc.PropertyDiff, error) {
+	baseDiff := makeBaseDiff(old, new, oldOk, newOk)
+	if baseDiff != Undecided {
+		etf, ok := etf.(shim.Schema)
+		if !ok {
+			etf = nil
+		}
+		propDiff := baseDiffToPropertyDiff(baseDiff, etf, eps)
+		if propDiff == nil {
+			return nil, nil
+		}
+		if mapHasReplacements(diff) {
+			propDiff = promoteToReplace(propDiff)
+		}
+		return map[string]*pulumirpc.PropertyDiff{string(key): propDiff}, nil
+	}
+	return nil, errors.New("diff is not simplified")
+}
+
+func propertyDiffResult(etf shim.Schema, eps *SchemaInfo, diff *pulumirpc.PropertyDiff) *pulumirpc.PropertyDiff {
+	// See pkg/cross-tests/diff_cross_test.go
+	// TestAttributeCollectionForceNew, TestBlockCollectionForceNew, TestBlockCollectionElementForceNew
+	// for a full case study of replacements in TF
+	if isForceNew(etf, eps) {
+		return promoteToReplace(diff)
+	}
+	return diff
+}
+
+func makeTopPropDiff(
+	old, new resource.PropertyValue,
+	oldOk, newOk bool,
+	etf shim.Schema,
+	eps *SchemaInfo,
+) *pulumirpc.PropertyDiff {
+	baseDiff := makeBaseDiff(old, new, oldOk, newOk)
+	if baseDiff != Undecided {
+		return baseDiffToPropertyDiff(baseDiff, etf, eps)
+	}
+
+	if !old.DeepEquals(new) {
+		return propertyDiffResult(etf, eps, &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE})
+	}
+	return nil
+}
+
+func makePropDiff(
+	ctx context.Context,
+	key resource.PropertyKey,
+	etf shim.Schema,
+	eps *SchemaInfo,
+	old, new resource.PropertyValue,
+	oldOk, newOk bool,
+) map[string]*pulumirpc.PropertyDiff {
```

Comment:
Should it short-circuit if `!oldOk || !newOk`? or it still needs to traverse everything to compute force new perhaps.

Reply:
Yeah, exactly - we need to traverse everything to calculate ForceNew